File: src/engine/opening.js
import { STARTING_GOLD } from "./economy.js";
import { pushCaptured } from "./capturesEvents.js";

/* ============================================================================
 * Opening helpers
 * - first-turn decision
 * - initial player template
 * - opening board/hand normalization
 * ========================================================================== */

/* 1) First-turn decision */
export function decideFirstTurn(humanCard, aiCard, rng = Math.random) {
  const hour = new Date().getHours();
  const isNight = hour >= 18 || hour < 6;

  let winnerKey = "human";
  let reason = "";
  if (humanCard.month === aiCard.month) {
    winnerKey = rng() < 0.5 ? "human" : "ai";
    reason = "same month random";
  } else if (isNight) {
    winnerKey = humanCard.month < aiCard.month ? "human" : "ai";
    reason = "night rule: lower month starts";
  } else {
    winnerKey = humanCard.month > aiCard.month ? "human" : "ai";
    reason = "day rule: higher month starts";
  }

  const winnerLabel = winnerKey === "human" ? "Player" : "AI";
  const log =
    `Starter decided [${isNight ? "night" : "day"}]: ` +
    `Player ${humanCard.month} vs AI ${aiCard.month} -> ${winnerLabel} (${reason})`;

  return { winnerKey, log };
}

/* 2) Player template */
export function emptyPlayer(label) {
  return {
    label,
    hand: [],
    captured: { kwang: [], ribbon: [], five: [], junk: [] },
    gukjinMode: "five",
    gukjinLocked: false,
    goCount: 0,
    turnCount: 0,
    lastGoBase: 0,
    presidentHold: false,
    presidentHoldMonth: null,
    shakingDeclaredMonths: [],
    heldBonusCards: [],
    ppukState: {
      active: false,
      streak: 0,
      lastTurnNo: 0,
      lastSource: null,
      lastMonth: null
    },
    gold: STARTING_GOLD,
    declaredStop: false,
    score: 0,
    events: {
      pansseul: 0,
      ppuk: 0,
      jjob: 0,
      shaking: 0,
      bomb: 0,
      ddadak: 0,
      ssul: 0,
      jabbeok: 0,
      yeonPpuk: 0
    }
  };
}

/* 3) Opening normalization */
export function normalizeOpeningHands(players, remain, initLog) {
  // Opening bonus cards are not auto-triggered in hand phase.
  // The effect is applied only when the player actually acquires the card.
  return remain.slice();
}

export function normalizeOpeningBoard(board, remain, firstPlayer, initLog) {
  let nextBoard = board.slice();
  let nextRemain = remain.slice();
  let changed = true;
  while (changed) {
    changed = false;
    const bonus = nextBoard.filter((c) => c.bonus?.stealPi);
    if (bonus.length === 0) break;
    changed = true;
    bonus.forEach((card) => {
      pushCaptured(firstPlayer.captured, card);
      initLog.push(`Opening adjust: ${firstPlayer.label} captures board bonus card ${card.name}`);
    });
    nextBoard = nextBoard.filter((c) => !c.bonus?.stealPi);
    while (nextBoard.length < 8 && nextRemain.length > 0) {
      nextBoard.push(nextRemain[0]);
      nextRemain = nextRemain.slice(1);
    }
  }
  return { board: nextBoard, remain: nextRemain };
}

/* 4) Opening special checks */
export function findPresidentMonth(cards) {
  const counts = {};
  for (const card of cards) {
    // President check only counts normal month cards (1..12).
    // Excludes pass/dummy (month 0) and bonus cards (month 13).
    if (!card || card.passCard) continue;
    if (card.month < 1 || card.month > 12) continue;
    counts[card.month] = (counts[card.month] || 0) + 1;
    if (counts[card.month] >= 4) return card.month;
  }
  return null;
}

File: experiments/k_hyperneat_matgo/py_eval/opening.py
from __future__ import annotations

from datetime import datetime
from typing import Callable

from .captures import push_captured
from .economy import STARTING_GOLD
from .models import Card, PlayerState


def decide_first_turn(human_card: Card, ai_card: Card, rng: Callable[[], float]) -> dict[str, str]:
    hour = datetime.now().hour
    is_night = hour >= 18 or hour < 6

    winner_key = "human"
    reason = ""
    if human_card.month == ai_card.month:
        winner_key = "human" if rng() < 0.5 else "ai"
        reason = "same month random"
    elif is_night:
        winner_key = "human" if human_card.month < ai_card.month else "ai"
        reason = "night rule: lower month starts"
    else:
        winner_key = "human" if human_card.month > ai_card.month else "ai"
        reason = "day rule: higher month starts"

    winner_label = "Player" if winner_key == "human" else "AI"
    log = (
        f"Starter decided [{'night' if is_night else 'day'}]: "
        f"Player {human_card.month} vs AI {ai_card.month} -> {winner_label} ({reason})"
    )
    return {"winnerKey": winner_key, "log": log}


def empty_player(label: str) -> PlayerState:
    return PlayerState(label=label, gold=STARTING_GOLD)


def normalize_opening_hands(players: dict[str, PlayerState], remain: list[Card], init_log: list[str]) -> list[Card]:
    _ = (players, init_log)
    return list(remain)


def normalize_opening_board(board: list[Card], remain: list[Card], first_player: PlayerState, init_log: list[str]) -> dict[str, list[Card]]:
    next_board = list(board)
    next_remain = list(remain)
    changed = True
    while changed:
        changed = False
        bonus_cards = [card for card in next_board if card.bonus and card.bonus.get("stealPi")]
        if not bonus_cards:
            break
        changed = True
        for card in bonus_cards:
            push_captured(first_player.captured, card)
            init_log.append(f"Opening adjust: {first_player.label} captures board bonus card {card.name}")
        next_board = [card for card in next_board if not (card.bonus and card.bonus.get("stealPi"))]
        while len(next_board) < 8 and next_remain:
            next_board.append(next_remain.pop(0))
    return {"board": next_board, "remain": next_remain}


def find_president_month(cards: list[Card]) -> int | None:
    counts: dict[int, int] = {}
    for card in cards:
        if not card:
            continue
        if getattr(card, "pass_card", False):
            continue
        if card.month < 1 or card.month > 12:
            continue
        counts[card.month] = counts.get(card.month, 0) + 1
        if counts[card.month] >= 4:
            return card.month
    return None
